osd: allow remote read by calling cls method from within cls context

kiizawa opened the pull request:
This commit implements cls_cxx_gather() function, which can be used by a cls method to gather data from multiple source objects in parallel.

Fixes: https://tracker.ceph.com/issues/48182
Signed-off-by: Ken Iizawa <iizawa.ken@fujitsu.com>

ofriedma commented:
@kiizawa I'm not sure but it looks like there is no option to fetch only partial data (offset and length) from the object like in cls_cxx_read2, What do you think about adding this option too?

athanatos commented:
What happens if there is an interval change while the operation is waiting on the remote reads?

kiizawa commented:
@ofriedma

Right, you cannot directly execute cls_cxx_read2 on the remote OSD.

If you want to do so, you need to implement a cls method that execute cls_cxx_read2 locally,
and then execute the cls method on the remote OSD.

For now, I do not plan to make this PR generic enough to be able to execute all read APIs including cls_cxx_read2 on the remote OSD.

kiizawa commented:
@athanatos

I did some experiments to see what happens.

I killed the source OSD while the target OSD is waiting for the completion of the remote read from the source OSD.
In most cases, the target OSD successfully completes the remote read by automatically retrying the new primary.

However, it turns out that the target OSD aborts when the target OSD is the new primary.
For example, let's say osd.5 (the target OSD) is reading src_object from osd.2 (the source OSD).
src_object belongs to a PG consisting of three OSDs [2,5,3].
When I kill osd.2 during the remote read, osd.5 becomes the new primary, and aborts when retrying the new primary (osd.5 itself).

I have been struggling to figure out what causes the abort.
I would appreciate your advice if you could find possible causes from the stack trace shown bellow.

```
2020-12-08T19:43:32.874+0900 7fb460f98700 -1 *** Caught signal (Aborted) **
 in thread 7fb460f98700 thread_name:tp_osd_tp

 ceph version 16.0.0-7217-g8d4a8735c3 (8d4a8735c3a13ec1583475c8524b19a33d1369f6) pacific (dev)
 1: /home/ceph/build/bin/ceph-osd(+0x30962e2) [0x556803bd22e2]
 2: /lib64/libpthread.so.0(+0x12dd0) [0x7fb48595cdd0]
 3: gsignal()
 4: abort()
 5: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x379) [0x556803c808cd]
 6: /home/ceph/build/bin/ceph-osd(+0x3144980) [0x556803c80980]
 7: (PrimaryLogPG::on_flushed()+0x204) [0x5568033cd4f0]
 8: (PeeringState::complete_flush()+0x52) [0x55680363c490]
 9: (FlushState::~FlushState()+0x7a) [0x5568032edaf4]
 10: (void __gnu_cxx::new_allocator<FlushState>::destroy<FlushState>(FlushState*)+0x1c) [0x556803338954]
 11: (void std::allocator_traits<std::allocator<FlushState> >::destroy<FlushState>(std::allocator<FlushState>&, FlushState*)+0x23) [0x556803338215]
 12: (std::_Sp_counted_ptr_inplace<FlushState, std::allocator<FlushState>, (__gnu_cxx::_Lock_policy)2>::_M_dispose()+0x37) [0x55680333759f]
 13: (std::_Sp_counted_base<(__gnu_cxx::_Lock_policy)2>::_M_release()+0x42) [0x5568030ea2f4]
 14: (std::__shared_count<(__gnu_cxx::_Lock_policy)2>::~__shared_count()+0x27) [0x5568030e6821]
 15: (std::__shared_ptr<FlushState, (__gnu_cxx::_Lock_policy)2>::~__shared_ptr()+0x1c) [0x5568032edb40]
 16: (std::shared_ptr<FlushState>::~shared_ptr()+0x18) [0x5568032edb5c]
 17: (ContainerContext<std::shared_ptr<FlushState> >::~ContainerContext()+0x2e) [0x5568033369ba]
 18: (ContainerContext<std::shared_ptr<FlushState> >::~ContainerContext()+0x18) [0x5568033369e2]
 19: (Context::complete(int)+0x45) [0x55680317fb95]
 20: (OSD::ShardedOpWQ::handle_oncommits(std::__cxx11::list<Context*, std::allocator<Context*> >&)+0x8d) [0x55680319651b]
 21: (OSD::ShardedOpWQ::_process(unsigned int, ceph::heartbeat_handle_d*)+0x953) [0x55680316e1b5]
 22: (ShardedThreadPool::shardedthreadpool_worker(unsigned int)+0x59a) [0x556803c6bb34]
 23: (ShardedThreadPool::WorkThreadSharded::entry()+0x25) [0x556803c6d749]
 24: (Thread::entry_wrapper()+0x83) [0x556803c57be7]
 25: (Thread::_entry_func(void*)+0x18) [0x556803c57b5a]
 26: /lib64/libpthread.so.0(+0x82de) [0x7fb4859522de]
 27: clone()
```

athanatos commented:
What debugging have you tried so far?  Have you enabled osd debugging (debug osd = 20 debug ms = 1)?

athanatos commented:
Also, if you go a bit higher in that log you should be able to see which assert actually failed.

athanatos commented:
Ah, I think it failed the

    ceph_assert(object_contexts.empty());

assert.

The problem is that cls_gather returns -EINPROGRESS while attaching the OpContext (via ptr) to the C_gather callback.  In the event that the objecter read completes successfully, this is fine as the callback will run execute_ctx which will complete the request and (in close_op_ctx) delete the OpContext releasing the ObjectContextRef references held.  However, if an interval change happens in the mean time, we flush outstanding IO and then assert that no ObjectContextRefs are still live (the above assert).  What needs to happen is that upon interval change (PrimaryLogPG::on_change is probably the right place) any outstanding cls_gather operations need to have their state cleaned up (OpContext deleted, op requeued).  See PrimaryLogPG::cancel_flush_ops -> cancel_flush for an example.  You'll likely need to add a structure to PrimaryLogPG tracking any such ops.

kiizawa commented:
@athanatos

I really appreciate your help. Yes, that's the assert failing.
I'll implement the clean-up according to your advice.

kiizawa commented:
@athanatos

I have implemented the clean-up on interval change.
The assert does not fail now.

Could you take a look ?

athanatos commented on src/osd/PrimaryLogPG.cc:
```diff
@@ -10500,6 +10507,35 @@ bool PrimaryLogPG::is_present_clone(hobject_t coid)
+    dout(0) << __func__ << " " << cgop->obc->obs.oi.soid << " tid " << *p << dendl;
+  }
+  cgop->objecter_tids.clear();
+  object_contexts.purge(cgop->obc->obs.oi.soid);
```
Why?

kiizawa commented:
@athanatos

I was not aware of cancel_cls_gather() not deleting OpContext,
which in turn was preventing object_contexts from getting cleared,
so I tried to clear object_contexts in a wrong way.
Now cancel_cls_gather() deletes OpContext and object_contexts gets cleared.

I also fixed the logging level.

athanatos commented on src/osd/PrimaryLogPG.cc:
```diff
@@ -9952,6 +9954,10 @@ int PrimaryLogPG::cls_gather(OpContext *ctx, std::map<std::string, bufferlist*>
+  CLSGatherOpRef cgop(std::make_shared<CLSGatherOp>());
+  cgop->ctx = ctx;
+  cgop->obc = obc;
+  cgop->op = op;
```
Add a constructor to CLSGatherOp

kiizawa commented:
@athanatos

I added a constructor to CLSGatherOp.

athanatos commented on src/osd/PrimaryLogPG.cc:
```diff
@@ -9924,6 +9924,60 @@ void PrimaryLogPG::cancel_copy_ops(bool requeue, vector<ceph_tid_t> *tids)
+      return;
+    // TODO: check errors
+    pg->lock();
+    pg->cls_gather_ops.erase(oid);
```
I think here you need to check whether the op is still in cls_gather_ops.  If not, the op was necessarily cancelled and you should drop the lock and bail.  objecter cancellation is meant to free resources, and won't necessarily prevent that race.

athanatos commented on src/osd/PrimaryLogPG.cc:
```diff
@@ -9924,6 +9924,60 @@ void PrimaryLogPG::cancel_copy_ops(bool requeue, vector<ceph_tid_t> *tids)
+  CLSGatherOpRef cgop(std::make_shared<CLSGatherOp>(ctx, obc, op));
+  for (std::map<std::string, bufferlist*>::iterator it = src_objs->begin(); it != src_objs->end(); it++) {
+    std::string oid = it->first;
+    it->second = new bufferlist;
```
Can src_objs simply be a std::map<string, bufferlist>?  I don't really think we want to do allocation here and trust that the caller will correctly free it.

athanatos replied:
Actually, the user should probably pass in const std::set<std::string> & and internally we should construct a std::map<std::string, std::bufferlist> to be returned via std::move in cls_cxx_get_gathered_data.  That would entirely avoid reference ownership ambiguity.

athanatos commented on src/osd/PrimaryLogPG.h:
```diff
@@ -256,6 +257,19 @@ class PrimaryLogPG : public PG, public PGBackend::Listener {
+      : ctx(ctx_), obc(obc_), op(op_) {}
+    ~CLSGatherOp() {}
+  };
+  typedef std::shared_ptr<CLSGatherOp> CLSGatherOpRef;
```
using =

kiizawa replied:
@athanatos

Could you explain this comment a little more, please?
I don't quite understand what I am supposed to do.

athanatos replied:
We generally use C++ syntax for this sort of thing.  Replace this line with:

```using CLSGatherOpRef = std::shared_ptr<CLSGatherOp>;```

However, note below that I'm not really sure why this needs to be a shared_ptr in the first place.

athanatos commented on src/osd/PrimaryLogPG.cc:
```diff
@@ -9924,6 +9924,60 @@ void PrimaryLogPG::cancel_copy_ops(bool requeue, vector<ceph_tid_t> *tids)
+  ObjectContextRef obc = get_object_context(soid, false);
+  C_GatherBuilder gather(cct);
+
+  CLSGatherOpRef cgop(std::make_shared<CLSGatherOp>(ctx, obc, op));
```
auto =

athanatos replied:
Also, I don't entirely get why CLSGatherOpRef needs to be a shared_pointer.  I see only one reference owner -- cls_gather_ops, can cls_gather_ops directly contain CLSGatherOp by value?

athanatos commented:
Importing PrimaryLogPG.h into class_api.cc will break crimson builds.  Instead, you'll need to emulate cls_cxx_read2 -- add an extern declaration in objclass.h for your two interface mechanisms and move the implementations to objclass.cc -- it already includes PrimaryLogPG.h.  You'll also need to add stubs to crimson/osd/objclass.cc, but they don't need to work beyond satisfying the linker.

athanatos commented:
I apologize for not looking more closely before!

kiizawa commented:
@athanatos 

I have fixed all the problems you pointed out.

However, I noticed that C_gather::finish ignores and discards the return value of remote-reads unless it is -ECANCELED.
I am trying to modify source code to save the return value somewhere and pass it to the the caller of cls_cxx_get_gathered_data.

kiizawa commented:
@athanatos

I modified cls_cxx_get_gathered_data to return the error code of remote-reads.
I also did a rebase.

athanatos commented on src/osd/PrimaryLogPG.cc:
```diff
@@ -10078,6 +10078,76 @@ void PrimaryLogPG::cancel_copy_ops(bool requeue, vector<ceph_tid_t> *tids)
+  ObjectContextRef obc = get_object_context(soid, false);
+  C_GatherBuilder gather(cct);
+
+  auto cgop = CLSGatherOp(ctx, obc, op);
```
auto &cgop = cls_gather_ops[soid];

athanatos replied:
You should be able to avoid the second copy.

kiizawa replied:
@athanatos 

I tried the followings and somehow got compilation errors in both cases.

auto [iter, inserted] = cls_gather_ops.emplace(soid, {ctx, obc, op});
auto [iter, inserted] = cls_gather_ops.emplace({soid, {ctx, obc, op}});

Instead, I am calling the constructor of CLSGatherOp explicitly.

athanatos commented on src/osd/PrimaryLogPG.cc:
```diff
@@ -10870,6 +10940,43 @@ bool PrimaryLogPG::is_present_clone(hobject_t coid)
+  p->second.rval = r;
+}
+
+void PrimaryLogPG::cancel_cls_gather(CLSGatherOp cgop, bool requeue,
```
Probably pass an interator into cls_gather_ops to avoid the extra lookup.

athanatos commented on src/osd/PrimaryLogPG.cc:
```diff
@@ -10870,6 +10940,43 @@ bool PrimaryLogPG::is_present_clone(hobject_t coid)
+  dout(10) << __func__ << dendl;
+  map<hobject_t,CLSGatherOp>::iterator p = cls_gather_ops.begin();
+  while (p != cls_gather_ops.end()) {
+    cancel_cls_gather((p++)->second, requeue, tids);
```
Here, you may as well pass in p++.

athanatos commented on src/osd/objclass.cc:
```diff
@@ -700,3 +700,44 @@ uint64_t cls_get_pool_stripe_width(cls_method_context_t hctx)
+    results->clear();
+  } else {
+    GatherFinisher *gf = (GatherFinisher*)op_finisher;
+    *results = *gf->src_obj_buffs;
```
swap?

athanatos replied:
Looks like this hasn't been addressed?

athanatos replied:
Also, guard with a null check on results.  When using a pointer out param like results here, we normally permit a nullptr to indicate that the return should be discarded.

athanatos commented:
I'll finish looking at this later this evening.

athanatos commented on src/osd/PrimaryLogPG.cc:
```diff
@@ -10078,6 +10078,76 @@ void PrimaryLogPG::cancel_copy_ops(bool requeue, vector<ceph_tid_t> *tids)
   }
 }
 
+struct C_gather : public Context {
```
You also need to record the current peering interval.  See C_CopyFrom::last_peering_reset.  Upon construction, use pg->get_last_peering_reset() to get the start epoch of the interval and check against it in finish.

athanatos commented on src/osd/PrimaryLogPG.cc:
```diff
@@ -10078,6 +10078,76 @@ void PrimaryLogPG::cancel_copy_ops(bool requeue, vector<ceph_tid_t> *tids)
+  void finish(int r) override {
+    if (r == -ECANCELED)
+      return;
+    pg->lock();
```
Use std::scoped_lock l{*pg}; to avoid having to unlock explicitly.

athanatos commented on src/osd/PrimaryLogPG.cc:
```diff
@@ -10078,6 +10078,76 @@ void PrimaryLogPG::cancel_copy_ops(bool requeue, vector<ceph_tid_t> *tids)
+  map<hobject_t,PrimaryLogPG::CLSGatherOp>::iterator p = cls_gather_ops.find(soid);
+  ceph_assert(p != cls_gather_ops.end());
+  int r = p->second.rval;
+  cls_gather_ops.erase(soid);
```
erase(p)

athanatos commented on src/osd/PrimaryLogPG.cc:
```diff
@@ -10078,6 +10078,76 @@ void PrimaryLogPG::cancel_copy_ops(bool requeue, vector<ceph_tid_t> *tids)
+  }
+};
+
+int PrimaryLogPG::start_cls_gather(OpContext *ctx, std::shared_ptr<std::map<std::string, bufferlist> > src_obj_buffs, const std::string& pool,
```
I don't really understand why src_obj_buffs is a shared_ptr.  Once this call completes, the only reference is held by the finisher in the caller.

athanatos commented:
I've finished commenting on this version. @kiizawa

kiizawa commented:
@athanatos

I made changes according to your comments.

sebastian-philipp commented:
jenkins test make check

athanatos commented on src/osd/objclass.cc:
```diff
@@ -700,3 +700,45 @@ uint64_t cls_get_pool_stripe_width(cls_method_context_t hctx)
+    GatherFinisher *gf = (GatherFinisher*)op_finisher;
+    *results = *gf->src_obj_buffs;
+    delete gf->src_obj_buffs;
+    r = (*pctx)->pg->finish_cls_gather(*pctx);
```
This is the bit that actually cleans up the gather state in the PG, right?  What if the class making use of this interface fails to call cls_cxx_get_gathered_data?  I think you need to call this in the C_gather callback.

athanatos commented on src/osd/objclass.cc:
```diff
@@ -700,3 +700,45 @@ uint64_t cls_get_pool_stripe_width(cls_method_context_t hctx)
+		   const char *cls, const char *method, bufferlist& inbl)
+{
+  PrimaryLogPG::OpContext **pctx = (PrimaryLogPG::OpContext**)hctx;
+  std::map<std::string, bufferlist> *src_obj_buffs = new std::map<std::string, bufferlist>;
```
We really try not to do direct allocations like this as it can lead to leaks, like this one does.  GatherFinisher doesn't actually free src_obj_buffs in its destructor, so unless cls_cxx_get_gathered_data runs, we leak the buffers.  Modify GatherFinisher to directly contain src_obj_buffs by value and populate it directly here.  Something like:

```
auto [iter, inserted] = (*pctx->op_finishers).emplace(std::make_pair((*pctx)->current_osd_subop_num, std::make_unique<GatherFinisher>());
assert(inserted);
auto &bufs = iter->second;
for (const auto &obj : src_objs) {
  (bufs)[obj];
}
```

athanatos replied:
The OpContext owns the reference to the GatherFinisher via the unique_ptr and the GatherFinisher owns the buffers directly ensuring that the destruction of the OpContext through whatever means will clean up the buffers.

athanatos commented on src/osd/objclass.cc:
```diff
@@ -700,3 +700,45 @@ uint64_t cls_get_pool_stripe_width(cls_method_context_t hctx)
 }
+
+struct GatherFinisher : public PrimaryLogPG::OpFinisher {
+  std::map<std::string, bufferlist> *src_obj_buffs;
```
Make this a value and remove it from the constructor.

athanatos commented on src/osd/objclass.cc:
```diff
@@ -700,3 +700,45 @@ uint64_t cls_get_pool_stripe_width(cls_method_context_t hctx)
+  } else {
+    GatherFinisher *gf = (GatherFinisher*)op_finisher;
+    *results = *gf->src_obj_buffs;
+    delete gf->src_obj_buffs;
```
This should be unnecessary with the other changes.

athanatos commented:
@kiizawa A bit more to clean up, but I think we're getting close.

kiizawa commented:
@athanatos 

I fixed the potential memory leak you pointed out.

I also noticed that CLSGatherOp is never erased from cls_gather_ops unless cls_cxx_get_gathered_data runs.
This is because in ea97316 I modified cls_cxx_get_gathered_data to read the return value kept in CLSGatherOp.

I moved the return value from CLSGatherOp to OpContext so that C_gather can erase CLSGatherOp from cls_gather_ops.

kiizawa commented:
I fixed a Signed-off-by error.

kiizawa commented:
jenkins retest this please

athanatos commented:
jenkins test make check

athanatos commented on src/osd/PrimaryLogPG.cc:
```diff
@@ -10078,6 +10078,70 @@ void PrimaryLogPG::cancel_copy_ops(bool requeue, vector<ceph_tid_t> *tids)
+struct C_gather : public Context {
+  PrimaryLogPGRef pg;
+  hobject_t oid;
+  PrimaryLogPG::OpContext *ctx;
```
How does this get cleaned up if we don't run execute_ctx?

athanatos replied:
I think the answer is that it gets cleared in cancel_cls_gather as the same pointer is captured in the cls_gather_ops map.  Please remove this member here and access if from cls_gather_ops -- you already have to do the lookup below.

athanatos commented on src/osd/PrimaryLogPG.cc:
```diff
@@ -10078,6 +10078,70 @@ void PrimaryLogPG::cancel_copy_ops(bool requeue, vector<ceph_tid_t> *tids)
+    if (r == -ECANCELED)
+      return;
+    std::scoped_lock locker{*pg};
+    map<hobject_t,PrimaryLogPG::CLSGatherOp>::iterator p = pg->cls_gather_ops.find(oid);
```
auto p =

athanatos commented on src/osd/objclass.cc:
```diff
@@ -700,3 +700,48 @@ uint64_t cls_get_pool_stripe_width(cls_method_context_t hctx)
+    results->clear();
+  } else {
+    GatherFinisher *gf = (GatherFinisher*)op_finisher;
+    *results = gf->src_obj_buffs;
```
*results = std::move(gf->src_obj_buffs);

athanatos commented on src/osd/objclass.cc:
```diff
@@ -700,3 +700,48 @@ uint64_t cls_get_pool_stripe_width(cls_method_context_t hctx)
+      op_finisher = op_finisher_it->second.get();
+    }
+  }
+  if (op_finisher == NULL) {
```
nullptr

athanatos commented:
Ok, other than the comments above, I think we're good.  The only other thing we need is at least a basic test.  I think the easiest thing to do would be to add a method to cls/cls_sdk with a corresponding test in test/cls_sdk/test_cls_sdk.cc.  @neha-ojha Does that seem reasonable for new cls sdk features?

kiizawa commented:
@athanatos

Done.

I will add a method to cls/cls_sdk with a corresponding test in test/cls_sdk/test_cls_sdk.cc, if that is OK with @neha-ojha.

neha-ojha commented:
@athanatos @kiizawa `cls/sdk/cls_sdk.cc` only uses the api exposed by `src/include/rados/objclass.h`, this PR is adding a method in `src/objclass/objclass.h` so that don't work.

athanatos commented:
I see.  In that case, I'd suggest adding an object class to invoke this method along with a test to exercise it.

kiizawa commented:
@athanatos

I added an object class and a corresponding test.

I think I should also test the case where an interval change happens,
but I do not know how to do it inside Ceph's test framework.

I would appreciate it if you could give me some advice.

kiizawa commented:
jenkins retest this please

kiizawa commented:
jenkins test make check

kiizawa commented:
@athanatos

I noticed that start_cls_gather() was not using the specified pool name for remote-reads.
I fixed it.

athanatos commented:
@kiizawa If you add this test to qa/workunits/rados/test.sh, it'll get run during failure tests in the normal test suite -- that's probably the way to go.

athanatos commented on src/cls/remote_reads/cls_remote_reads.cc:
```diff
@@ -0,0 +1,87 @@
+/*
```
Probably ought to call this cls_test_remote_reads?

athanatos commented on src/osd/PrimaryLogPG.h:
```diff
@@ -257,6 +258,19 @@ class PrimaryLogPG : public PG, public PGBackend::Listener {
   typedef std::shared_ptr<FlushOp> FlushOpRef;
 
+  struct CLSGatherOp {
+    OpContext *ctx;
```
Initialize this to nullptr.

athanatos commented on src/osd/PrimaryLogPG.h:
```diff
@@ -257,6 +258,19 @@ class PrimaryLogPG : public PG, public PGBackend::Listener {
+    ObjectContextRef obc;
+    OpRequestRef op;
+    std::vector<ceph_tid_t> objecter_tids;
+    int rval;
```
Initialize this to something.

athanatos commented on src/osd/objclass.cc:
```diff
@@ -700,3 +700,48 @@ uint64_t cls_get_pool_stripe_width(cls_method_context_t hctx)
+}
+
+int cls_cxx_get_gathered_data(cls_method_context_t hctx, std::map<std::string, bufferlist> *results)
+{
```
assert(results) or check below for null.

kiizawa commented:
@athanatos 

I have finished all except for adding the test to qa/workunits/rados/test.sh.

It seems that tests run by qa/workunits/rados/test.sh need to be prefixed by 'ceph_test_rados_'.
Should I rename the test 'ceph_test_cls_test_remote_reads' to 'ceph_test_rados_cls_test_remote_reads', or use other workarounds such as creating a symbolic link from 'ceph_test_cls_test_remote_reads' to 'ceph_test_rados_cls_test_remote_reads'?

Do you have any recommendations?

athanatos commented:
@kiizawa Renaming would seem to be the path of least resistance.

kiizawa commented:
@athanatos

Now the test can be run by qa/workunits/rados/test.sh.

athanatos commented:
Close, let's put the test in test/librados/ though for consistency.

kiizawa commented:
@athanatos

Done.

athanatos reviewed (approved):
LGTM pending test run.

athanatos commented:
Looks like it hasn't been picked up for a test run, I'll kick on off this evening/tomorrow.

tchaikov commented:
@athanatos I will take care of this. just missed it.

athanatos commented:
@tchaikov Oh, thanks!

tchaikov commented on src/test/librados/cls_remote_reads.cc:
```diff
@@ -0,0 +1,52 @@
+
+using namespace librados;
+
+TEST(ClsTestRemoteReads, TestGather) {
```
```
2021-03-25T20:39:37.253 INFO:tasks.workunit.client.0.smithi166.stdout:     api_cls_remote_reads: [ RUN      ] ClsTestRemoteReads.TestGather
2021-03-25T20:39:37.254 INFO:tasks.workunit.client.0.smithi166.stdout:     api_cls_remote_reads: /home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/17.0.0-2494-g51f534a
d/rpm/el8/BUILD/ceph-17.0.0-2494-g51f534ad/src/test/librados/cls_remote_reads.cc:46: Failure
2021-03-25T20:39:37.254 INFO:tasks.workunit.client.0.smithi166.stdout:     api_cls_remote_reads: Expected equality of these values:
2021-03-25T20:39:37.254 INFO:tasks.workunit.client.0.smithi166.stdout:     api_cls_remote_reads:   0
2021-03-25T20:39:37.254 INFO:tasks.workunit.client.0.smithi166.stdout:     api_cls_remote_reads:   ioctx.exec("tgt_object", "test_remote_reads", "test_gather", in, out)
2021-03-25T20:39:37.254 INFO:tasks.workunit.client.0.smithi166.stdout:     api_cls_remote_reads:     Which is: -1
2021-03-25T20:39:37.255 INFO:tasks.workunit.client.0.smithi166.stdout:     api_cls_remote_reads: [  FAILED  ] ClsTestRemoteReads.TestGather (7437 ms)
2021-03-25T20:39:37.255 INFO:tasks.workunit.client.0.smithi166.stdout:     api_cls_remote_reads: [----------] 1 test from ClsTestRemoteReads (7437 ms total)
2021-03-25T20:39:37.255 INFO:tasks.workunit.client.0.smithi166.stdout:     api_cls_remote_reads:
2021-03-25T20:39:37.255 INFO:tasks.workunit.client.0.smithi166.stdout:     api_cls_remote_reads: [----------] Global test environment tear-down
2021-03-25T20:39:37.255 INFO:tasks.workunit.client.0.smithi166.stdout:     api_cls_remote_reads: [==========] 1 test from 1 test suite ran. (7437 ms total)
2021-03-25T20:39:37.256 INFO:tasks.workunit.client.0.smithi166.stdout:     api_cls_remote_reads: [  PASSED  ] 0 tests.
2021-03-25T20:39:37.256 INFO:tasks.workunit.client.0.smithi166.stdout:     api_cls_remote_reads: [  FAILED  ] 1 test, listed below:
2021-03-25T20:39:37.256 INFO:tasks.workunit.client.0.smithi166.stdout:     api_cls_remote_reads: [  FAILED  ] ClsTestRemoteReads.TestGather
```

see https://pulpito.ceph.com/kchai-2021-03-25_14:56:28-rados-wip-kefu-testing-2021-03-25-2042-distro-basic-smithi/5998005/

athanatos replied:
@kiizawa Looks like the new tests failed?  The above link should give you the relevant logs.

kiizawa replied:
@athanatos, @tchaikov 

I took a look at the OSD logs.
It seems like OSDs failed to load libcls_test_remote_reads.so due to a permission problem.

> 2021-03-25T20:39:37.195+0000 7f952c0be700 20 osd.5 pg_epoch: 68 pg[24.2( empty local-lis/les=67/68 n=0 ec=67/67 lis/c=67/67 les/c/f=68/68/0 sis=67) [5,4,0] r=0 lpr=67 crt=0'0 mlcod 0'0 active+clean] do_op: op osd_op(client.24703.0:5 24.2 24:41780a1c:::tgt_object:head [call test_remote_reads.test_gather in=237b] snapc 0=[] ondisk+read+known_if_redirected e68) v8
>
> 2021-03-25T20:39:37.195+0000 7f952c0be700  0 _get_class not permitted to load test_remote_reads

Would it be possible to run the test with the following options?

> [osd]
> osd class load list = *
> osd class default list = *

Alternatively, I could modify src/common/options.cc so that these lists include "test_remote_reads" by default.

athanatos replied:
The rados_api_tests.yaml file can provide osd config overrides.  I think what you want is probably:

```
diff --git a/qa/suites/rados/basic/tasks/rados_api_tests.yaml b/qa/suites/rados/basic/tasks/rados_api_tests.yaml
index a54e03b1a9c..a1b136f7cc2 100644
--- a/qa/suites/rados/basic/tasks/rados_api_tests.yaml
+++ b/qa/suites/rados/basic/tasks/rados_api_tests.yaml
@@ -16,6 +16,9 @@ overrides:
         debug ms: 1
       mon:
         mon warn on pool no app: false
+      osd:
+        osd class load list: *
+        osd class default list: *
 tasks:
 - workunit:
     clients:
```

athanatos replied:
I'm a bit sceptical about exposing this as a default class as its purpose is really for testing.

athanatos replied:
@kiizawa ^

athanatos replied:
Btw, there's a brand new trio of video guides on how to use/develop teuthology tests if you are interested.  https://youtu.be/3ZYBMGfQKVk is the first one.

kiizawa replied:
@athanatos

I modified the rados_api_tests.yaml file.

Thanks, I will watch the video.

tchaikov commented on qa/suites/rados/basic/tasks/rados_api_tests.yaml:
```diff
@@ -16,6 +16,9 @@ overrides:
       mon:
         mon warn on pool no app: false
+      osd:
+        osd class load list: *
```
this is not a valid YAML.  as `*` is used to start an anchor name. see https://yaml.org/spec/1.2/spec.html#*%20alias//

tchaikov replied:
might want to quote `*` like
```yaml
osd:
  osd class load list: "*"
```

kiizawa commented:
@tchaikov 

Thanks, I fixed it.

tchaikov commented:
@kiizawa thanks. could you please squash the commit of "qa: fix invalid YAML format" into the commit which introduced the change of `osd class load list: *`?

kiizawa commented:
@tchaikov 

Done.

tchaikov commented on qa/workunits/cls/test_cls_test_remote_reads.sh:
```diff
@@ -0,0 +1,5 @@
+#!/bin/sh -e
+
+ceph_test_rados_cls_test_remote_reads
```
```
2021-04-03T01:34:18.923 INFO:tasks.workunit:Running workunit cls/test_cls_test_remote_reads.sh...
2021-04-03T01:34:18.924 DEBUG:teuthology.orchestra.run.gibba011:workunit test cls/test_cls_test_remote_reads.sh> mkdir -p -- /home/ubuntu/cephtest/mnt.0/client.0/tmp && cd -- /home/ubuntu/cephtest/mnt.0/client.0/tmp && CEPH_CLI_TEST_DUP_COMMAND=1 CEPH_REF=3d0f85a6c3046372f708a5f2893231db573d2f75 TESTDIR="/home/ubuntu/cephtest" CEPH_ARGS="--cluster ceph" CEPH_ID="0" PATH=$PATH:/usr/sbin CEPH_BASE=/home/ubuntu/cephtest/clone.client.0 CEPH_ROOT=/home/ubuntu/cephtest/clone.client.0 adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage timeout 3h /home/ubuntu/cephtest/clone.client.0/qa/workunits/cls/test_cls_test_remote_reads.sh
2021-04-03T01:34:18.978 DEBUG:teuthology.orchestra.run:got remote process result: 127
2021-04-03T01:34:18.978 INFO:tasks.workunit.client.0.gibba011.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cls/test_cls_test_remote_reads.sh: line 3: ceph_test_rados_cls_test_remote_reads: command not found
2021-04-03T01:34:18.979 INFO:tasks.workunit:Stopping ['cls'] on client.0...
```

kiizawa commented:
@tchaikov 

Sorry for taking up your time.

qa/workunits/cls/test_cls_test_remote_reads.sh is an obsolete test I should have deleted. 
My test is run by qa/workunits/rados/test.sh instead.

tchaikov commented:
@kiizawa no worries. that's why we are testing the changes =)

could you please squash the new commit into the one creating `test_cls_test_remote_reads.sh` ?

kiizawa commented:
@tchaikov 

I squashed 7 test-related commits, including the latest commit, into the one creating ```test_cls_test_remote_reads.sh```.

kiizawa commented:
jenkins test make check

tchaikov commented:
the test failure should be fixed by #40579.

tchaikov commented:
jenkins test make check

tchaikov commented on src/test/librados/cls_remote_reads.cc:
```diff
@@ -0,0 +1,52 @@
+
+using namespace librados;
+
+TEST(ClsTestRemoteReads, TestGather) {
```
```
2021-04-05T20:26:13.260 INFO:tasks.workunit.client.0.gibba021.stdout:     api_cls_remote_reads: [ RUN      ] ClsTestRemoteReads.TestGather
2021-04-05T20:26:13.260 INFO:tasks.workunit.client.0.gibba021.stdout:     api_cls_remote_reads: /home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/17.0.0-2843-gc52566ed
/rpm/el8/BUILD/ceph-17.0.0-2843-gc52566ed/src/test/librados/cls_remote_reads.cc:46: Failure
2021-04-05T20:26:13.261 INFO:tasks.workunit.client.0.gibba021.stdout:     api_cls_remote_reads: Expected equality of these values:
2021-04-05T20:26:13.261 INFO:tasks.workunit.client.0.gibba021.stdout:     api_cls_remote_reads:   0
2021-04-05T20:26:13.262 INFO:tasks.workunit.client.0.gibba021.stdout:     api_cls_remote_reads:   ioctx.exec("tgt_object", "test_remote_reads", "test_gather", in, out)
2021-04-05T20:26:13.262 INFO:tasks.workunit.client.0.gibba021.stdout:     api_cls_remote_reads:     Which is: -1
2021-04-05T20:26:13.262 INFO:tasks.workunit.client.0.gibba021.stdout:     api_cls_remote_reads: [  FAILED  ] ClsTestRemoteReads.TestGather (1372 ms)
```

https://pulpito.ceph.com/kchai-2021-04-05_10:46:36-rados-wip-kefu-testing-2021-04-05-1650-distro-basic-gibba/6021968/

tchaikov reviewed (changes requested):
ClsTestRemoteReads.TestGather fails.

kiizawa commented:
@tchaikov, @athanatos 

Thanks!

I took a look at ```ceph-osd.7.log```.

```
2021-04-05T20:26:12.983+0000 7f49fb749700 20 osd.7 pg_epoch: 19 pg[19.2( empty local-lis/les=18/19 n=0 ec=18/18 lis/c=18/18 les/c/f=19/19/0 sis=18) [7,4] r=0 lpr=18 crt=0'0 mlcod 0'0 active+clean] do_op: op osd_op(client.4563.0:5 19.2 19:41780a1c:::tgt_object:head [call test_remote_reads.test_gather in=236b] snapc 0=[] ondisk+read+known_if_redirected e19) v8
2021-04-05T20:26:12.983+0000 7f49fb749700  0 _get_class not permitted to load test_remote_reads
2021-04-05T20:26:12.983+0000 7f49fb749700  1 -- [v2:172.21.2.130:6824/30195,v1:172.21.2.130:6825/30195] --> 172.21.2.121:0/4169045365 -- osd_op_reply(5 tgt_object [call] v0'0 uv0 ondisk = -1 ((1) Operation not permitted)) v8 -- 0x564278e41b00 con 0x564278fed800
```

This looks like the same error that happend in the first run,
which means the following settings on ```qa/suites/rados/basic/tasks/rados_api_tests.yaml``` are not working.


```
      osd:
        osd class load list: "*"
        osd class default list: "*"
```

Does anyone see what I am missing?

I guess the above settings should show up on ```teuthology.log```,
but I don't see them.

```
  overrides:
    admin_socket:
      branch: wip-kefu-testing-2021-04-05-1650
    ceph:
      conf:
        client:
          debug monc: 20
          debug ms: 1
          debug objecter: 20
          debug rados: 20
        global:
          lockdep: true
          mon client directed command retry: 5
          mon election default strategy: 1
          ms inject socket failures: 5000
          ms type: async
        mgr:
          debug mgr: 20
          debug ms: 1
        mon:
          debug mon: 20
          debug ms: 1
          debug paxos: 20
          mon min osdmap epochs: 50
          mon osdmap full prune interval: 2
          mon osdmap full prune min: 15
          mon osdmap full prune txsize: 2
          mon scrub interval: 300
          mon warn on pool no app: false
          paxos service trim min: 10
        osd:
          bluestore block size: 96636764160
          bluestore compression algorithm: zstd
          bluestore compression mode: aggressive
          bluestore fsck on mount: true
          debug bluefs: 20
          debug bluestore: 20
          debug monc: 20
          debug ms: 1
          debug osd: 20
          debug rocksdb: 10
          mon osd backfillfull_ratio: 0.85
          mon osd full ratio: 0.9
          mon osd nearfull ratio: 0.8
          osd debug verify cached snaps: true
          osd debug verify missing on start: true
          osd failsafe full ratio: 0.95
          osd objectstore: bluestore
          osd op queue: debug_random
          osd op queue cut off: debug_random
          osd shutdown pgref assert: true
      fs: xfs
      log-ignorelist:
      - \(MDS_ALL_DOWN\)
      - \(MDS_UP_LESS_THAN_MAX\)
      - \(OSD_SLOW_PING_TIME
      - reached quota
      - overall HEALTH_
      - \(CACHE_POOL_NO_HIT_SET\)
      - \(POOL_FULL\)
      - \(SMALLER_PGP_NUM\)
      - \(SLOW_OPS\)
      - \(CACHE_POOL_NEAR_FULL\)
      - \(POOL_APP_NOT_ENABLED\)
      - \(PG_AVAILABILITY\)
      - \(OBJECT_MISPLACED\)
      - slow request
      log-whitelist:
      - \(MDS_ALL_DOWN\)
      - \(MDS_UP_LESS_THAN_MAX\)
      sha1: c52566ed0f013e5ec0b6153e7f14c1560d36f673
    ceph-deploy:
      conf:
        client:
          log file: /var/log/ceph/ceph-$name.$pid.log
        mon:
          osd default pool size: 2
    install:
      ceph:
        sha1: c52566ed0f013e5ec0b6153e7f14c1560d36f673
    selinux:
      whitelist:
      - scontext=system_u:system_r:logrotate_t:s0
    thrashosds:
      bdev_inject_crash: 2
      bdev_inject_crash_probability: 0.5
    workunit:
      branch: wip-kefu-testing-2021-04-05-1650
      sha1: c52566ed0f013e5ec0b6153e7f14c1560d36f673
```

tchaikov commented on qa/workunits/rados/test.sh:
```diff
@@ -29,6 +29,7 @@ for f in \
     api_service api_service_pp \
     api_c_write_operations \
     api_c_read_operations \
+    api_cls_remote_reads \
```
this work unit is picked up by more than one task yaml file. `qa/suites/rados/basic/tasks/rados_api_tests.yaml` is but one of them. the failed test was
```
rados/verify/{centos_latest ceph clusters/{fixed-2 openstack} d-thrash/none mon_election/classic msgr-failures/few msgr/async objectstore/bluestore-comp-zstd rados tasks/rados_api_tests validater/lockdep}
```

where `rados/verify/tasks/rados_api_tests` is used as one of the workloads. i'd suggest
```sh
git grep -w rados/test.sh | grep suites
```
for finding all the yml file which uses the changed work unit. but there are couple other ways to address this issue

- probably to update the settings right in `qa/workunits/rados/test.sh`.
- add a new .sh file exercising `api_cls_remote_reads` in `qa/workunits/cls/. because `rados_cls_all.yaml` already takes care of this by enabling all osd classes.

kiizawa commented:
@tchaikov

Thank you for your detailed explanation.

> add a new .sh file exercising api_cls_remote_reads in qa/workunits/cls/. because rados_cls_all.yaml` already takes care of this by enabling all osd classes.

I would like my test to get run by ```qa/workunits/rados/test.sh``` as I and @athanatos talked in https://github.com/ceph/ceph/pull/38120#issuecomment-791022360.

> probably to update the settings right in qa/workunits/rados/test.sh.

I also tried adding the following lines to qa/workunits/rados/test.sh, but it did not work.

```
 declare -A pids

+ceph config set osd osd_class_load_list "*"
+ceph config set osd osd_class_default_list "*"
+
 for f in \
     api_aio api_aio_pp \
     api_io api_io_pp \
```

Thus, modifying all the YAML files found by ```git grep -w rados/test.sh | grep suites``` seems to be the only solution.
I have finished modifying them and squashed the commit to the commit that modified ```qa/suites/rados/basic/tasks/rados_api_tests.yaml```.

tchaikov commented:
@kiizawa @athanatos i'd prefer the 2nd option -- to add a test under `qa/workunits/cls/`. but i don't feel strong either way.

athanatos commented:
I can't think of a reason not to globally enable osd classes across all teuthology tests.  @tchaikov ?

tchaikov commented:
@athanatos that's a good idea also. we can do this in a follow-up PR. IIRC, we do have a template .conf file which is used for all tests.

tchaikov commented:
- https://pulpito.ceph.com/kchai-2021-04-11_12:36:03-rados-wip-kefu-testing-2021-04-11-1826-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-04-12_02:29:32-rados-wip-kefu-testing-2021-04-12-0057-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-04-12_08:01:43-rados-wip-kefu-testing-2021-04-12-1256-distro-basic-smithi/

multi-site failure tracked by https://tracker.ceph.com/issues/50218
